### PR TITLE
[fga] project visibility and prebuild perms

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Introduction of project visibilities:
 - private
 - org-public (default)
 - public

This change doesn't add any UI. Projects have `org-public` visibility. 
I have added this [issue](https://linear.app/gitpod/issue/EXP-471/project-visibility) for follow-up work, such as adding a settings option in the UI.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 14f1884</samp>

This pull request adds and improves features for project visibility and prebuild permissions, using the SpiceDB service and schema. It updates the `Authorizer`, `ProjectsService`, and `GitpodServerImpl` classes, as well as the `gitpod-protocol` and `spicedb` modules, to handle the new visibility and permission logic. It also adds tests and fixes some formatting issues.

</details>

## Related Issue(s)

Fixes EXP-468

## How to test

1) Start a workspace using https://se-project-vis.preview.gitpod-dev.com/#https://github.com/svenefftinge/2048-in-terminal
There is a project with a prebuild for this repo in my org, but your workspace should not get one.

I.e. you should **not!!!** see this message in the first terminal:
Screenshot 2023-08-18 at 10 40 58

2) Join my org (https://se-project-vis.preview.gitpod-dev.com/orgs/join?inviteId=0b8713e6-97f4-4a9d-9836-06344d23ab38) and create another workspace on https://se-project-vis.preview.gitpod-dev.com/#https://github.com/svenefftinge/2048-in-terminal
Now you should get a workspace with a prebuild.

I.e. you should see this message in the first terminal:
Screenshot 2023-08-18 at 10 40 58

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
        <li><b>🏷️ Name</b> - se-project-vis</li>
        <li><b>🔗 URL</b> - <a href="https://se-project-vis.preview.gitpod-dev.com/workspaces" target="_blank">se-project-vis.preview.gitpod-dev.com/workspaces</a>.</li>
        <li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
        <li><b>📦 Version</b> - </li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by jeanp413** [comment]
Don't see any project after joining your org :thinking: 
image

**Comment by geropl** [comment]
Starting to review now :eyes:

File: components/spicedb/schema/schema.yaml [hunk 1]
```diff
@@ -86,14 +87,17 @@ schema: |-
     // A subject is a viewer, if:
     //  * the user is directly assigned as a viewer
     //  * the project has granted access to everyone in an organization
-    relation viewer: user | organization#member
+    relation viewer: user | organization#member | user:*
```
Comment (geropl) [inline components/spicedb/schema/schema.yaml:90]:
```suggestion
    //  * the project has granted access to _any_ user on this installation
    relation viewer: user | organization#member | user:*
```

File: components/spicedb/codegen/codegen.go [hunk 2]
```diff
@@ -158,20 +162,38 @@ func generateFluentAPI(def *corev1.NamespaceDefinition) string {
 
 		objects := ``
 		for _, t := range rel.TypeInformation.AllowedDirectRelations {
-			objects += `
+			if t.GetPublicWildcard() != nil {
```
Comment (geropl) [inline components/spicedb/codegen/codegen.go:165]: Nice, need the same in the "workpsace sharing" PR, and did not yet look into it, yet. :orange_heart:

File: components/server/src/projects/projects-service.ts [hunk 3]
```diff
@@ -239,9 +239,11 @@ export class ProjectsService {
                 await db.storeProject(project);
 
                 await this.auth.addProjectToOrg(installer.id, teamId, project.id);
+                await this.auth.setProjectVisibility(installer.id, project.id, teamId, "org-public");
             });
         } catch (err) {
             await this.auth.removeProjectFromOrg(installer.id, teamId, project.id);
+            await this.auth.setProjectVisibility(installer.id, project.id, teamId, "private");
```
Comment (geropl) [inline components/server/src/projects/projects-service.ts:246]: The PR description says project visibility is "org-public" is the default, but here it's "private"? I think this is bc we want to remove all relations, but might warrant a comment (at least for future me :upside_down:)

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 4]
```diff
@@ -394,6 +394,13 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
                 return;
             }
 
+            // check if the user has access to the project
+            if (
```
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:398]: Do you think we also need to check this inside WorkspaceService.startWorkspace to guard the access [here](https://github.com/gitpod-io/gitpod/blob/d2b220f11cee9ac07688ee802fafe6c8dcc80c75/components/server/src/workspace/workspace-factory.ts#L55-L59)? :thinking:

File: components/server/src/projects/projects-service.ts (hunk 3 above)
Comment (svenefftinge) [reply components/server/src/projects/projects-service.ts:246]: Yes. I can also make the removal part of removeProjectFromOrg which seems better

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 4 above)
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:398]: I'll check, but I wonder how a prebuild is relevant when starting an existing workspace.

**Comment by geropl** [comment]
I like the schema changes, and the code looks all good to me! (did not test, yet).

There is one thing we should talk through in sync I feel: If we have dedicated permissions for read/writing prebuilds (which I think is good!): How do we decide if the user also has access to the underlying workspace (of type `"prebuild"`) and workspaceInstance?

In the past, the permission checks for `Workspace`, it's `instance` and `Prebuild` were always identical for a simple: Our API requires us to have access to all three in a lot of spots, because simply accessing `Prebuild` is not enough for most use-cases (it does not contain all infos from the instance+ws): Examples are: listing prebuilds, or watch logs of a Prebuild

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 4 above)
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:398]: Sorry, I meant WorkspaceService.**create**Workspace :roll_eyes:

**Comment by svenefftinge** [comment]
I'm redeploying the preview env to reproduce what jean pierre shared above ☝️

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works :heavy_check_mark:

**Pull request merged**